File: lib/features/rankings/models/ranking_models.dart
import 'package:clashkingapp/core/constants/image_assets.dart';
import 'package:clashkingapp/core/config/api_config.dart';

enum RankingAudience { players, clans }

enum RankingPeriod { current, history }

enum RankingSource { official, clashKing }

enum RankingBoard {
  playerHome(
    audience: RankingAudience.players,
    source: RankingSource.official,
    supportsLocation: true,
    supportsHistory: true,
    iconUrl: ImageAssets.trophies,
  ),
  playerBuilder(
    audience: RankingAudience.players,
    source: RankingSource.official,
    supportsLocation: true,
    supportsHistory: true,
    iconUrl: ImageAssets.builderBaseTrophy,
  ),
  playerTownHall(
    audience: RankingAudience.players,
    source: RankingSource.clashKing,
    supportsLocation: false,
    supportsHistory: true,
    iconUrl: ImageAssets.trophies,
  ),
  playerRanked(
    audience: RankingAudience.players,
    source: RankingSource.clashKing,
    supportsLocation: false,
    supportsHistory: true,
    iconUrl: ImageAssets.legendLeagueOne,
  ),
  clanHome(
    audience: RankingAudience.clans,
    source: RankingSource.official,
    supportsLocation: true,
    supportsHistory: true,
    iconUrl: ImageAssets.trophies,
  ),
  clanBuilder(
    audience: RankingAudience.clans,
    source: RankingSource.official,
    supportsLocation: true,
    supportsHistory: true,
    iconUrl: ImageAssets.builderBaseTrophy,
  ),
  clanCapital(
    audience: RankingAudience.clans,
    source: RankingSource.official,
    supportsLocation: true,
    supportsHistory: true,
    iconUrl: ImageAssets.capitalTrophy,
  ),
  clanDonations(
    audience: RankingAudience.clans,
    source: RankingSource.clashKing,
    supportsLocation: true,
    supportsHistory: false,
    supportsWorldwide: false,
    iconUrl: ImageAssets.clanGamesMedals,
  ),
  clanWarWins(
    audience: RankingAudience.clans,
    source: RankingSource.clashKing,
    supportsLocation: true,
    supportsHistory: false,
    supportsWorldwide: false,
    iconUrl: ImageAssets.war,
  ),
  clanWinStreak(
    audience: RankingAudience.clans,
    source: RankingSource.clashKing,
    supportsLocation: false,
    supportsHistory: false,
    iconUrl: ImageAssets.war,
  );

  const RankingBoard({
    required this.audience,
    required this.source,
    required this.supportsLocation,
    required this.supportsHistory,
    required this.iconUrl,
    this.supportsWorldwide = true,
  });

  final RankingAudience audience;
  final RankingSource source;
  final bool supportsLocation;
  final bool supportsHistory;
  final bool supportsWorldwide;
  final String iconUrl;

  bool get isClan => audience == RankingAudience.clans;
}

class RankingLocation {
  const RankingLocation({
    required this.id,
    required this.name,
    required this.isCountry,
    this.countryCode,
    this.isWorldwide = false,
  });

  const RankingLocation.worldwide()
    : id = null,
      name = 'Worldwide',
      isCountry = false,
      countryCode = null,
      isWorldwide = true;

  final int? id;
  final String name;
  final bool isCountry;
  final String? countryCode;
  final bool isWorldwide;

  String get apiPath => isWorldwide ? 'global' : id.toString();

  bool get hasValidCountryCode =>
      isCountry && RegExp(r'^[A-Za-z]{2}$').hasMatch(countryCode ?? '');

  factory RankingLocation.fromJson(Map<String, dynamic> json) {
    return RankingLocation(
      id: _asIntOrNull(json['id']),
      name: json['name']?.toString().trim() ?? '',
      isCountry: json['isCountry'] == true,
      countryCode: json['countryCode']?.toString().trim().toUpperCase(),
    );
  }

  @override
  bool operator ==(Object other) =>
      other is RankingLocation &&
      other.id == id &&
      other.isWorldwide == isWorldwide;

  @override
  int get hashCode => Object.hash(id, isWorldwide);
}

class RankingLeagueOption {
  const RankingLeagueOption({
    required this.id,
    required this.name,
    required this.iconUrl,
  });

  static const legendOne = RankingLeagueOption(
    id: 105000036,
    name: 'Legend I',
    iconUrl: ImageAssets.legendLeagueOne,
  );

  final int id;
  final String name;
  final String iconUrl;
}

class RankingQuery {
  const RankingQuery({
    required this.board,
    required this.location,
    required this.period,
    required this.historyDate,
    required this.townHallLevel,
    required this.leagueTier,
  });

  final RankingBoard board;
  final RankingLocation location;
  final RankingPeriod period;
  final DateTime historyDate;
  final int townHallLevel;
  final RankingLeagueOption leagueTier;
}

class RankingResult {
  const RankingResult({
    required this.entries,
    required this.source,
    required this.limit,
  });

  final List<RankingEntry> entries;
  final RankingSource source;
  final int limit;
}

class RankingEntry {
  const RankingEntry({
    required this.audience,
    required this.rank,
    required this.previousRank,
    required this.tag,
    required this.name,
    required this.subtitle,
    required this.score,
    required this.imageUrl,
    required this.metricImageUrl,
    required this.townHallLevel,
    this.clanBadgeUrl = '',
  });

  final RankingAudience audience;
  final int rank;
  final int previousRank;
  final String tag;
  final String name;
  final String subtitle;
  final int score;
  final String imageUrl;
  final String metricImageUrl;
  final int townHallLevel;
  final String clanBadgeUrl;

  String get movement {
    if (previousRank <= 0 || rank <= 0) return '=';
    final delta = previousRank - rank;
    if (delta == 0) return '=';
    return delta > 0 ? '+$delta' : '$delta';
  }

  factory RankingEntry.fromJson(
    Map<String, dynamic> json,
    RankingBoard board, {
    String? rankedLeagueIconUrl,
  }) {
    final tag = _firstString(json, const ['tag', 'player_tag', 'clan_tag']);
    final townHall = _firstInt(json, const [
      'townHallLevel',
      'townhall_level',
      'townhallLevel',
    ]);
    final clanName =
        _nestedString(json['clan'], 'name') ??
        _firstString(json, const ['clan_name', 'clanName']);
    final clanTag =
        _nestedString(json['clan'], 'tag') ??
        _firstString(json, const ['clan_tag', 'clanTag']);
    final subtitleParts = <String>[
      if (clanName.isNotEmpty) clanName,
      if (clanTag.isNotEmpty && clanTag != tag) clanTag,
      if (board.isClan && clanName.isEmpty && clanTag.isEmpty && tag.isNotEmpty)
        tag,
    ];
    final includedClanBadge =
        _nestedString(json['clan'], 'badge') ??
        _nestedString(json['clan'], 'badgeUrls.medium') ??
        _nestedString(json['clan'], 'badge_urls.medium') ??
        _firstString(json, const ['clan_badge', 'clanBadge']);
    final clanBadgeUrl = board.isClan || clanTag.isEmpty
        ? ''
        : board.source == RankingSource.official
        ? '${ApiConfig.apiUrlV2}/clan/${Uri.encodeComponent(clanTag)}/badge'
        : includedClanBadge;

    final leagueIcon =
        _nestedString(json['leagueTier'], 'iconUrls.medium') ??
        _nestedString(json['leagueTier'], 'iconUrls.large') ??
        _nestedString(json['leagueTier'], 'iconUrls.small') ??
        _nestedString(json['leagueTier'], 'badge') ??
        _nestedString(json['league'], 'iconUrls.medium') ??
        _nestedString(json['league'], 'iconUrls.large') ??
        _nestedString(json['league'], 'iconUrls.small') ??
        _nestedString(json['league'], 'badge');
    final badgeUrl =
        _nestedString(json['badgeUrls'], 'medium') ??
        _nestedString(json['badge_urls'], 'medium') ??
        _firstString(json, const ['badge_url']);
    final selectedRankedLeagueIcon = board == RankingBoard.playerRanked
        ? (rankedLeagueIconUrl ?? board.iconUrl)
        : null;
    final playerImageUrl = townHall > 0
        ? ImageAssets.townHall(townHall)
        : (selectedRankedLeagueIcon ?? leagueIcon ?? board.iconUrl);
    final imageUrl = board.isClan
        ? (badgeUrl.isEmpty ? ImageAssets.clanCastle : badgeUrl)
        : playerImageUrl;
    final metricImageUrl = board.isClan
        ? board.iconUrl
        : (selectedRankedLeagueIcon ?? leagueIcon ?? board.iconUrl);

    return RankingEntry(
      audience: board.audience,
      rank: _firstInt(json, const ['rank', 'placement']),
      previousRank: _firstInt(json, const ['previousRank', 'previous_rank']),
      tag: tag,
      name: _firstString(json, const ['name', 'player_name', 'clan_name'], tag),
      subtitle: subtitleParts.join(' · '),
      score: _scoreFor(json, board),
      imageUrl: imageUrl,
      metricImageUrl: metricImageUrl,
      townHallLevel: townHall,
      clanBadgeUrl: clanBadgeUrl,
    );
  }
}

int _scoreFor(Map<String, dynamic> json, RankingBoard board) {
  final keys = switch (board) {
    RankingBoard.playerBuilder => const [
      'builderBaseTrophies',
      'builder_base_trophies',
      'versusTrophies',
      'trophies',
    ],
    RankingBoard.playerRanked => const [
      'league_trophies',
      'leagueTrophies',
      'trophies',
    ],
    RankingBoard.clanHome => const ['clanPoints', 'clan_points'],
    RankingBoard.clanBuilder => const [
      'clanBuilderBasePoints',
      'clanVersusPoints',
      'clan_builder_base_points',
    ],
    RankingBoard.clanCapital => const [
      'clanCapitalPoints',
      'capitalPoints',
      'clan_capital_points',
    ],
    RankingBoard.clanDonations => const ['donations', 'troops_donated'],
    RankingBoard.clanWarWins => const ['war_wins', 'warWins'],
    RankingBoard.clanWinStreak => const ['war_win_streak', 'warWinStreak'],
    _ => const ['trophies'],
  };
  return _firstInt(json, keys);
}

int _firstInt(Map<String, dynamic> json, List<String> keys) {
  for (final key in keys) {
    final value = _asIntOrNull(json[key]);
    if (value != null) return value;
  }
  return 0;
}

int? _asIntOrNull(Object? value) {
  if (value is int) return value;
  if (value is num) return value.toInt();
  return int.tryParse(value?.toString() ?? '');
}

String _firstString(
  Map<String, dynamic> json,
  List<String> keys, [
  String fallback = '',
]) {
  for (final key in keys) {
    final value = json[key]?.toString().trim() ?? '';
    if (value.isNotEmpty) return value;
  }
  return fallback;
}

String? _nestedString(Object? raw, String path) {
  Object? current = raw;
  for (final segment in path.split('.')) {
    if (current is! Map) return null;
    current = current[segment];
  }
  final value = current?.toString().trim() ?? '';
  return value.isEmpty ? null : value;
}

File: expo/src/features/rankings/models/ranking-models.ts
import { ImageAssets } from '../../../core/assets/image-assets';

export const RankingAudience = {
  players: 'players',
  clans: 'clans',
} as const;
export type RankingAudienceValue = (typeof RankingAudience)[keyof typeof RankingAudience];

export const RankingPeriod = {
  current: 'current',
  history: 'history',
} as const;
export type RankingPeriodValue = (typeof RankingPeriod)[keyof typeof RankingPeriod];

export const RankingSource = {
  official: 'official',
  clashKing: 'clashKing',
} as const;
export type RankingSourceValue = (typeof RankingSource)[keyof typeof RankingSource];

export interface RankingBoardValue {
  readonly name: RankingBoardName;
  readonly audience: RankingAudienceValue;
  readonly source: RankingSourceValue;
  readonly supportsLocation: boolean;
  readonly supportsHistory: boolean;
  readonly supportsWorldwide: boolean;
  readonly iconUrl: string;
  readonly isClan: boolean;
}

function board(
  name: RankingBoardName,
  audience: RankingAudienceValue,
  source: RankingSourceValue,
  supportsLocation: boolean,
  supportsHistory: boolean,
  iconUrl: string,
  supportsWorldwide = true,
): RankingBoardValue {
  return Object.freeze({
    name,
    audience,
    source,
    supportsLocation,
    supportsHistory,
    supportsWorldwide,
    iconUrl,
    isClan: audience === RankingAudience.clans,
  });
}

export type RankingBoardName =
  | 'playerHome'
  | 'playerBuilder'
  | 'playerTownHall'
  | 'playerRanked'
  | 'clanHome'
  | 'clanBuilder'
  | 'clanCapital'
  | 'clanDonations'
  | 'clanWarWins'
  | 'clanWinStreak';

export const RankingBoard = {
  playerHome: board(
    'playerHome',
    RankingAudience.players,
    RankingSource.official,
    true,
    true,
    ImageAssets.trophies,
  ),
  playerBuilder: board(
    'playerBuilder',
    RankingAudience.players,
    RankingSource.official,
    true,
    true,
    ImageAssets.builderBaseTrophy,
  ),
  playerTownHall: board(
    'playerTownHall',
    RankingAudience.players,
    RankingSource.clashKing,
    false,
    false,
    ImageAssets.trophies,
  ),
  playerRanked: board(
    'playerRanked',
    RankingAudience.players,
    RankingSource.clashKing,
    false,
    false,
    ImageAssets.legendLeagueOne,
  ),
  clanHome: board(
    'clanHome',
    RankingAudience.clans,
    RankingSource.official,
    true,
    true,
    ImageAssets.trophies,
  ),
  clanBuilder: board(
    'clanBuilder',
    RankingAudience.clans,
    RankingSource.official,
    true,
    true,
    ImageAssets.builderBaseTrophy,
  ),
  clanCapital: board(
    'clanCapital',
    RankingAudience.clans,
    RankingSource.official,
    true,
    true,
    ImageAssets.capitalTrophy,
  ),
  clanDonations: board(
    'clanDonations',
    RankingAudience.clans,
    RankingSource.clashKing,
    true,
    false,
    ImageAssets.clanGamesMedals,
    false,
  ),
  clanWarWins: board(
    'clanWarWins',
    RankingAudience.clans,
    RankingSource.clashKing,
    true,
    false,
    ImageAssets.war,
    false,
  ),
  clanWinStreak: board(
    'clanWinStreak',
    RankingAudience.clans,
    RankingSource.clashKing,
    false,
    false,
    ImageAssets.war,
  ),
} as const;

export const rankingBoards = Object.values(RankingBoard);

export class RankingLocation {
  constructor(
    readonly id: number | null,
    readonly name: string,
    readonly isCountry: boolean,
    readonly countryCode: string | null = null,
    readonly isWorldwide = false,
  ) {}

  static worldwide(): RankingLocation {
    return new RankingLocation(null, 'Worldwide', false, null, true);
  }

  static fromJson(json: Record<string, unknown>): RankingLocation {
    return new RankingLocation(
      asIntOrNull(json.id),
      stringValue(json.name),
      json.isCountry === true,
      json.countryCode == null ? null : stringValue(json.countryCode).toUpperCase(),
    );
  }

  get apiPath(): string {
    return this.isWorldwide ? 'global' : String(this.id);
  }

  get hasValidCountryCode(): boolean {
    return this.isCountry && /^[A-Za-z]{2}$/.test(this.countryCode ?? '');
  }

  equals(other: RankingLocation): boolean {
    return this.id === other.id && this.isWorldwide === other.isWorldwide;
  }
}

export class RankingLeagueOption {
  static readonly legendOne = new RankingLeagueOption(
    105000036,
    'Legend League 1',
    ImageAssets.legendLeagueOne,
  );
  static readonly legendTwo = new RankingLeagueOption(
    105000035,
    'Legend League 2',
    ImageAssets.legendLeagueTwo,
  );
  static readonly legendThree = new RankingLeagueOption(
    105000034,
    'Legend League 3',
    ImageAssets.legendLeagueThree,
  );

  constructor(
    readonly id: number,
    readonly name: string,
    readonly iconUrl: string,
  ) {}
}

export interface RankingQuery {
  readonly board: RankingBoardValue;
  readonly location: RankingLocation;
  readonly period: RankingPeriodValue;
  readonly historyDate: Date;
  readonly townHallLevel: number;
  readonly leagueTier: RankingLeagueOption;
}

export class RankingResult {
  constructor(
    readonly entries: readonly RankingEntry[],
    readonly source: RankingSourceValue,
    readonly limit: number,
  ) {}
}

export class RankingEntry {
  constructor(
    readonly audience: RankingAudienceValue,
    readonly rank: number,
    readonly previousRank: number,
    readonly tag: string,
    readonly name: string,
    readonly subtitle: string,
    readonly score: number,
    readonly imageUrl: string,
    readonly metricImageUrl: string,
    readonly townHallLevel: number,
    readonly clanBadgeUrl = '',
  ) {}

  get movement(): string {
    if (this.previousRank <= 0 || this.rank <= 0) return '=';
    const delta = this.previousRank - this.rank;
    return delta === 0 ? '=' : delta > 0 ? `+${delta}` : String(delta);
  }

  static fromJson(
    json: Record<string, unknown>,
    rankingBoard: RankingBoardValue,
    rankedLeagueIconUrl?: string,
  ): RankingEntry {
    const tag = firstString(json, ['tag', 'player_tag', 'clan_tag']);
    const townHall = firstInt(json, ['townHallLevel', 'townhall_level', 'townhallLevel']);
    const clanName =
      nestedString(json.clan, 'name') ?? firstString(json, ['clan_name', 'clanName']);
    const clanTag = nestedString(json.clan, 'tag') ?? firstString(json, ['clan_tag', 'clanTag']);
    const subtitle = rankingBoard.isClan
      ? [
          ...(clanName ? [clanName] : []),
          ...(clanTag && clanTag !== tag ? [clanTag] : []),
          ...(!clanName && !clanTag && tag ? [tag] : []),
        ].join(' · ')
      : clanName;
    const includedClanBadge =
      nestedString(json.clan, 'badgeUrls.small') ??
      nestedString(json.clan, 'badge_urls.small') ??
      nestedString(json.clan, 'badgeUrls.medium') ??
      nestedString(json.clan, 'badge_urls.medium') ??
      nestedString(json.clan, 'badgeUrls.large') ??
      nestedString(json.clan, 'badge_urls.large') ??
      nestedString(json.clan, 'badge') ??
      firstString(json, ['clan_badge', 'clanBadge']);
    const clanBadgeUrl =
      rankingBoard.isClan || !clanTag
        ? ''
        : rankingBoard.source === RankingSource.official
          ? ImageAssets.clanBadgeForTag(clanTag)
          : includedClanBadge;
    const leagueIcon =
      nestedString(json.leagueTier, 'iconUrls.medium') ??
      nestedString(json.leagueTier, 'iconUrls.large') ??
      nestedString(json.leagueTier, 'iconUrls.small') ??
      nestedString(json.leagueTier, 'badge') ??
      nestedString(json.league, 'iconUrls.medium') ??
      nestedString(json.league, 'iconUrls.large') ??
      nestedString(json.league, 'iconUrls.small') ??
      nestedString(json.league, 'badge');
    const builderIcon =
      rankingBoard === RankingBoard.playerBuilder
        ? ImageAssets.getBuilderBaseLeagueImage(json.builderBaseLeague)
        : null;
    const badgeUrl =
      nestedString(json.badgeUrls, 'small') ??
      nestedString(json.badge_urls, 'small') ??
      nestedString(json.badgeUrls, 'medium') ??
      nestedString(json.badge_urls, 'medium') ??
      nestedString(json.badgeUrls, 'large') ??
      nestedString(json.badge_urls, 'large') ??
      firstString(json, ['badge_url']);
    const rankedIcon =
      rankingBoard === RankingBoard.playerRanked
        ? (rankedLeagueIconUrl ?? rankingBoard.iconUrl)
        : null;
    const playerImage =
      rankingBoard === RankingBoard.playerBuilder
        ? (builderIcon ?? rankingBoard.iconUrl)
        : townHall > 0
          ? ImageAssets.townHall(townHall)
          : (rankedIcon ?? leagueIcon ?? rankingBoard.iconUrl);
    return new RankingEntry(
      rankingBoard.audience,
      firstInt(json, ['rank', 'placement']),
      firstInt(json, ['previousRank', 'previous_rank']),
      tag,
      firstString(json, ['name', 'player_name', 'clan_name'], rankingBoard.isClan ? tag : ''),
      subtitle,
      scoreFor(json, rankingBoard),
      rankingBoard.isClan ? badgeUrl || ImageAssets.clanCastle : playerImage,
      rankingBoard.isClan
        ? rankingBoard.iconUrl
        : (rankedIcon ?? leagueIcon ?? rankingBoard.iconUrl),
      townHall,
      clanBadgeUrl,
    );
  }
}

function scoreFor(json: Record<string, unknown>, rankingBoard: RankingBoardValue): number {
  const keys =
    rankingBoard === RankingBoard.playerBuilder
      ? ['builderBaseTrophies', 'builder_base_trophies', 'versusTrophies', 'trophies']
      : rankingBoard === RankingBoard.playerRanked
        ? ['league_trophies', 'leagueTrophies', 'trophies']
        : rankingBoard === RankingBoard.clanHome
          ? ['clanPoints', 'clan_points']
          : rankingBoard === RankingBoard.clanBuilder
            ? [
                'clanBuilderBasePoints',
                'builderBasePoints',
                'clanVersusPoints',
                'clan_builder_base_points',
              ]
            : rankingBoard === RankingBoard.clanCapital
              ? ['clanCapitalPoints', 'capitalPoints', 'clan_capital_points']
              : rankingBoard === RankingBoard.clanDonations
                ? ['donations', 'troops_donated']
                : rankingBoard === RankingBoard.clanWarWins
                  ? ['war_wins', 'warWins']
                  : rankingBoard === RankingBoard.clanWinStreak
                    ? ['war_win_streak', 'warWinStreak']
                    : ['trophies'];
  return firstInt(json, keys);
}

function firstInt(json: Record<string, unknown>, keys: readonly string[]): number {
  for (const key of keys) {
    const value = asIntOrNull(json[key]);
    if (value !== null) return value;
  }
  return 0;
}

function asIntOrNull(value: unknown): number | null {
  if (typeof value === 'number' && Number.isFinite(value)) return Math.trunc(value);
  const text = String(value ?? '');
  if (!/^[+-]?\d+$/.test(text)) return null;
  const parsed = Number.parseInt(text, 10);
  return Number.isNaN(parsed) ? null : parsed;
}

function firstString(
  json: Record<string, unknown>,
  keys: readonly string[],
  fallback = '',
): string {
  for (const key of keys) {
    const value = stringValue(json[key]);
    if (value) return value;
  }
  return fallback;
}

function nestedString(raw: unknown, path: string): string | null {
  let current = raw;
  for (const segment of path.split('.')) {
    if (!isRecord(current)) return null;
    current = current[segment];
  }
  const value = stringValue(current);
  return value || null;
}

function stringValue(value: unknown): string {
  return value == null ? '' : String(value).trim();
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}
